### PR TITLE
[SPARK-17903][SQL] MetastoreRelation should talk to external catalog instead of hive client

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

`HiveExternalCatalog` should be the only interface to talk to the hive metastore. In `MetastoreRelation` we can just use `ExternalCatalog` instead of `HiveClient` to interact with hive metastore,  and add missing API in `ExternalCatalog`.
## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai @gatorsmile

**Comment by rxin** [comment]
cc @ericl

**Comment by ericl** [comment]
LGTM

**Comment by rxin** [comment]
A bunch of mysterious failures ...

**Comment by ericl** [comment]
LGTM if tests pass :)

**Comment by yhuai** [comment]
`Caused by: sbt.ForkMain$ForkError: org.apache.spark.sql.catalyst.analysis.NoSuchTableException: Table or view 'default' not found in database 'srcpart';` Seems we swapped the db and table name somehow.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 1]
```diff
@@ -636,6 +637,13 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     client.getPartitions(db, table, partialSpec)
   }
 
+  override def listPartitionsByFilter(
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: Need a test case in `HiveExternalCatalogSuite`, just like the other APIs?

**Comment by gatorsmile** [comment]
LGTM except one comment about test case

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: added

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Pull request closed without merging**